src/resolution/common.rs

```diff
@@ -49,18 +52,36 @@ pub enum NpmPackageVersionResolutionError {
 }
 
 #[derive(Debug, Clone)]
-pub struct NpmVersionResolver {
+pub struct NpmVersionResolver<'patch> {
   pub types_node_version_req: Option<VersionReq>,
+  pub patch_packages: &'patch HashMap<PackageName, Vec<NpmPackageVersionInfo>>,
 }
 
-impl NpmVersionResolver {
-  pub fn resolve_best_package_version_info<'info, 'version>(
-    &self,
+impl NpmVersionResolver<'_> {
+  pub fn resolve_best_package_version_info<'a, 'version>(
+    &'a self,
     version_req: &VersionReq,
-    package_info: &'info NpmPackageInfo,
+    package_info: &'a NpmPackageInfo,
     existing_versions: impl Iterator<Item = &'version Version>,
-  ) -> Result<&'info NpmPackageVersionInfo, NpmPackageVersionResolutionError>
-  {
+  ) -> Result<&'a NpmPackageVersionInfo, NpmPackageVersionResolutionError> {
+    // always attempt to resolve from the patched packages first
+    if let Some(version_infos) = self.patch_packages.get(&package_info.name) {
+      let mut best_version: Option<&'a NpmPackageVersionInfo> = None;
+      for version_info in version_infos {
+        let version = &version_info.version;
+        if version_req.matches(version) {
```

Comment:
will this panic if the version req is a tag?

Reply:
😭

Reply:
Ah, actually this will always be resolved at this point (our API for this really sucks... I wish we had a concept of maybe unresolved version reqs)

Reply:
Answer... yes. I misunderstood the code earlier: https://github.com/denoland/deno_npm/pull/99/files